### PR TITLE
Let measure behave more like on the web

### PR DESCRIPTION
**Pull request by woehrl01**
Nodes with a measure function needs to be measured even so it seems there is no available space. So it behaves more like on the web. Fix facebook/yoga#488

### CODE REVIEW

File: tests/YGMeasureTest.cpp [hunk 1]
```diff
@@ -179,6 +179,45 @@ TEST(YogaTest, dont_measure_when_min_equals_max_mixed_height_percent) {
...
 
+TEST(YogaTest, measure_nodes_on_at_most_with_zero_space)
+{
+  const YGNodeRef root = YGNodeNew();
+  YGNodeStyleSetHeight(root, 200);
+  YGNodeStyleSetFlexDirection(root, YGFlexDirectionColumn);
+  YGNodeStyleSetAlignItems(root, YGAlignStretch);
+  YGNodeStyleSetAlignContent(root, YGAlignStretch);
+  YGNodeStyleSetFlexGrow(root, 0);
+  YGNodeStyleSetOverflow(root, YGOverflowHidden);
```
Comment (emilsjolander) [inline tests/YGMeasureTest.cpp]: remove overflow line. Should have no effect on test

File: tests/YGMeasureTest.cpp [hunk 2]
```diff
@@ -179,6 +179,45 @@ TEST(YogaTest, dont_measure_when_min_equals_max_mixed_height_percent) {
...
+  YGNodeStyleSetAlignSelf(root, YGAlignAuto);
+  YGNodeStyleSetPositionType(root, YGPositionTypeRelative);
```
Comment (emilsjolander) [inline tests/YGMeasureTest.cpp]: relative is default, no need to set it

File: tests/YGMeasureTest.cpp [hunk 3]
```diff
@@ -179,6 +179,45 @@ TEST(YogaTest, dont_measure_when_min_equals_max_mixed_height_percent) {
   YGNodeFreeRecursive(root);
 }
 
+TEST(YogaTest, measure_nodes_on_at_most_with_zero_space)
+{
+  const YGNodeRef root = YGNodeNew();
+  YGNodeStyleSetHeight(root, 200);
+  YGNodeStyleSetFlexDirection(root, YGFlexDirectionColumn);
+  YGNodeStyleSetAlignItems(root, YGAlignStretch);
```
Comment (emilsjolander) [inline tests/YGMeasureTest.cpp]: stretch is default for align items. no need to set it

File: tests/YGMeasureTest.cpp [hunk 4]
```diff
@@ -179,6 +179,45 @@ TEST(YogaTest, dont_measure_when_min_equals_max_mixed_height_percent) {
...
+  YGNodeStyleSetAlignContent(root, YGAlignStretch);
+  YGNodeStyleSetFlexGrow(root, 0);
+  YGNodeStyleSetOverflow(root, YGOverflowHidden);
+  YGNodeStyleSetAlignSelf(root, YGAlignAuto);
+  YGNodeStyleSetPositionType(root, YGPositionTypeRelative);
+  YGNodeStyleSetJustifyContent(root, YGJustifyFlexStart);
```
Comment (emilsjolander) [inline tests/YGMeasureTest.cpp]: flex start is default for justify content. no need to set it

File: tests/YGMeasureTest.cpp [hunk 5]
```diff
@@ -179,6 +179,45 @@ TEST(YogaTest, dont_measure_when_min_equals_max_mixed_height_percent) {
...
+  int measureCount = 0;
+
+  const YGNodeRef root_child0 = YGNodeNew();
+  YGNodeStyleSetFlexDirection(root_child0, YGFlexDirectionColumn);
+  YGNodeStyleSetAlignItems(root_child0, YGAlignStretch);
+  YGNodeStyleSetAlignContent(root_child0, YGAlignStretch);
+  YGNodeStyleSetAlignSelf(root_child0, YGAlignStretch);
+  YGNodeStyleSetJustifyContent(root_child0, YGJustifyFlexStart);
+  YGNodeStyleSetOverflow(root_child0, YGOverflowHidden);
+  YGNodeStyleSetPadding(root_child0, YGEdgeLeft, 100);
```
Comment (emilsjolander) [inline tests/YGMeasureTest.cpp]: Use YGEdgeALL instead?

File: tests/YGMeasureTest.cpp (hunk 3 above)
Comment (woehrl01) [inline tests/YGMeasureTest.cpp]: You're right, I adopted all the values form the issue.

**Comment by woehrl01** [comment]
The test is not that obvious correct. The problem is: 
The parent height is 200 and the padding is 100 on both sides so the available size is 0, the measure size is non zero. But as we haven't constraint the other side, We should be still growable in the width dimension. 

In my opinion a measured node could still overflow it's parent, isn't it?

**Comment by emilsjolander** [comment]
Ok thanks for explaining. Growing the parent width makes sense in that case. Regarding overflow I don't know. I would look to web browsers for guidance on this.

**Comment by woehrl01** [comment]
According to [this fiddle](https://jsfiddle.net/uh9whoom/) it's overflowing.

**Comment by woehrl01** [comment]
Btw. I'll update the test and remove the default values.

**Comment by emilsjolander** [comment]
Why should this only handle at_most and not undefined?

**Comment by woehrl01** [comment]
You're right it should also measure on the undefined mode, too.

**Comment by woehrl01** [comment]
Sorry, I actually was wrong. We don't need to ```measure``` here, we just need to return the correct size in the "no available size" skipping. I moved the skipping to the calculation step to perform the same calculations. I just don't call ```measure``` but use a new ```gYGSizeZero``` for the measurement instead.

**Comment by emilsjolander** [comment]
Seems measure_zero_space_should_grow test is failing after the last update. Could you have a look?

**Comment by woehrl01** [comment]
It's failing as we don't measure anymore. I reduce the testcase for now. As enabling always measure would lead to a possible layoutbreaking change. This fixes for the current behaviour. We need to discuss how the final behaviour should be and possibly add those as some kind of flag?

**Comment by woehrl01** [comment]
I expanded the jsfiddle to some testcases which some of them definetly aren't rendered correct currently https://jsfiddle.net/uh9whoom/6/

**Comment by emilsjolander** [comment]
@woehrl01 As long as we are moving towards better web compatibility and not breaking any major existing use cases I'm happy. I would rather not introduce new flags unless absolutely necessary.

**Comment by emilsjolander** [comment]
This is causing some test failures on integration tests internally. I haven't tracked down why but it seems this causes nodes to be measured to larger sizes than before.

**Comment by woehrl01** [comment]
Yes, it returns now a different size, it now behaves from the size like a node without a measure function. If there isn't any space.

**Comment by woehrl01** [comment]
But if this is making a breaking change. Maybe we should directly move to a more web compatible behaviour. What do you think?

**Comment by emilsjolander** [comment]
> But if this is introducing a breaking change. Maybe we should directly move to a more web compatible behaviour. What do you think?

Yeah that sounds good, much easier to land if we only breaks things once.

**Comment by woehrl01** [comment]
Ok, but this will lead to that we will measure more often. I'll put up a bigger test suite and put up a PR. Could take me a bit 👍 .

**Comment by emilsjolander** [comment]
ok Thanks! Regarding increased measurements as long as it is in a couple edge cases that is fine. We will have a look once the code is up and i'll do some internal logging to figure out if we regressed # of measure calls in practice.

**Comment by emilsjolander** [comment]
Re-opening this as well. This caused fewer breakages but still some issues which I don't have time to resolve right away. I will try re-commiting this next week. Sorry about that :)
